src/documentation/execution-model.md

```diff
@@ -66,95 +62,126 @@ Instead of processing all elements simultaneously, the elements in a
 elements in the bundle must be retried (otherwise the entire pipeline fails),
-although they need not be retried with the same bundling.
+although they do not need to be retried with the same bundling.
 
-In the following illustration, two elements (in green) were successfully
-processed. The third element’s processing failed, and there are three elements
-(in yellow) still to be processed. We see that the same elements were retried
-and processing successfully completed. Note that as shown, the retry does not
-necessarily happen in the same worker thread as the original attempt.
+For this example, we will use the `ParDo` from figure 1 that has an input
+collection with nine elements and is divided into two bundles.
 
-Because we encountered a failure while processing an element in the input
-bundle, we had to reprocess all of the elements in the input bundle. Thus, the
-entire output bundle must be thrown away since all of the results it contains
-will be recomputed.
+In figure 6, the first worker successfully processes all five elements in bundle
+A. The second worker processes the four elements in bundle B: the first two
+elements were successfully processed, the third element’s processing failed, and
+there is one element still awaiting processing.
+
+We see that the runner retries all elements in bundle B and the processing
+completes successfully the second time. Note that the retry does not necessarily
+happen on the same worker as the original processing attempt, as shown in the
+diagram.
 
 ![failure_retry]({{ site.baseurl }}/images/execution_model_failure_retry.svg)
 
-If the failed transform is a `ParDo`, then the `DoFn` instance is torn down and
-abandoned.
+**Figure 6:** The processing of an element within bundle B fails, and another worker
+retries the entire bundle.
 
-### Coupled failure: Failures between transforms
+Because we encountered a failure while processing an element in the input
+bundle, we had to reprocess _all_ of the elements in the input bundle. This means
+the runner must throw away the entire output bundle since all of the results it
+contains will be recomputed.
+
+Note that if the failed transform is a `ParDo`, then the `DoFn` instance is torn
+down and abandoned.
+
+### Coupled failure: Failures between transforms {#coupled-failure}
 
 If a failure to process an element in `ParDo2` causes `ParDo1` to re-execute,
 these two steps are said to be _co-failing_.
 
-In the following illustration, two `ParDo` transforms are processing elements.
-While processing an element, the second `ParDo` fails (which is shown in red).
-As a result, the runner must discard and recompute the output of the second
-`ParDo`. Because the runner was executing the two `ParDo`s together, the output
-bundle from the first `ParDo` must also be thrown away, and the elements in the
-input bundle must be retried. These two `ParDo`s are co-failing.
+For this example, we will use the dependently parallel `ParDo`s from figure 4.
```

Comment:
I wouldn't call these "dependently parallel ParDos". We're trying to separate the terms, since two transforms may have coupled-failures without being dependently parallel.